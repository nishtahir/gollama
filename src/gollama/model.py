import json
from typing import Any


class OllamaManifest:
    def __init__(self, json: dict[str, Any]):
        self.json = json

    def config(self) -> dict:
        return self.json["config"]

    def layers(self) -> list[dict]:
        return self.json["layers"]

    def blobs(self) -> list[str]:
        blobs = []
        blobs.append(self.config()["digest"])

        for layer in self.layers():
            blobs.append(layer["digest"])
        return blobs

    @staticmethod
    def from_str(content: str) -> "OllamaManifest":
        return OllamaManifest(json.loads(content))

    @staticmethod
    def from_path(path: str) -> "OllamaManifest":
        with open(path, "r") as f:
            return OllamaManifest.from_str(f.read())


class OllamaModel:
    def __init__(self, name: str, version: str):
        self.name = name
        self.version = version

    def __str__(self):
        return f"{self.name}:{self.version}"

    @staticmethod
    def from_name(model_name: str) -> "OllamaModel":
        if ":" not in model_name:
            model_name = f"{model_name}:latest"

        name, version = model_name.split(":")
        if not _uses_custom_registry(name):
            # We need to determin if it's from the
            # ollama model registry. if there's no '/'
            # then the package is stored in the library
            # otherwise it's stored in the custom registry
            if "/" not in name:
                name = f"library/{name}"

            # now we can construct the fully qualified name
            # using the ollama model registry prefix
            name = f"registry.ollama.ai/{name}"

        return OllamaModel(name=name, version=version)

    @staticmethod
    def from_manifest_path(path: str) -> "OllamaModel":
        # models/manifests/registry.ollama.ai/library/smollm2/135m -> smollm2:135m
        name = path.split("/")[-2]
        version = path.split("/")[-1]
        return OllamaModel(name=name, version=version)


def _uses_custom_registry(image_name: str) -> bool:
    # We're borrowing some logic from docker to determine if the
    # image is from a custom registry. My understanding is that
    # docker assumes an image is from a custom registry if the
    # segment before the first '/' contains a '.' (dot), a ':' (colon),
    # or equals "localhost". Otherwise, it defaults to Docker Hub.
    parts = image_name.split("/", 1)

    if len(parts) == 1:
        return False

    registry_candidate = parts[0]

    return (
        ("." in registry_candidate)
        or (":" in registry_candidate)
        or (registry_candidate == "localhost")
    )
