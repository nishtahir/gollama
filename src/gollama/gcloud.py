import os

from google.cloud import storage
from tqdm import tqdm


def upload_blob(file_path: str, bucket_name: str, blob_name: str) -> None:
    storage_client = storage.Client()
    bucket = storage_client.bucket(bucket_name)
    blob = bucket.blob(blob_name)

    # take the file name from the blob name
    file_name = os.path.basename(blob_name)
    # remove the sha256- prefix
    if file_name.startswith("sha256-"):
        file_name = file_name.split("-")[1][:12]

    with open(file_path, "rb") as in_file:
        total_bytes = os.fstat(in_file.fileno()).st_size
        with tqdm.wrapattr(
            in_file,
            "read",
            total=total_bytes,
            miniters=1,
            desc=f"pushing {file_name}...",
        ) as file_obj:
            blob.upload_from_file(file_obj)


def download_blob(bucket_name: str, blob_name: str, file_path: str) -> None:
    storage_client = storage.Client()
    bucket = storage_client.bucket(bucket_name)
    blob = bucket.get_blob(blob_name)
    if blob is None:
        raise ValueError(f"Blob {blob_name} not found in bucket {bucket_name}")

    size = blob.size
    # take the file name from the blob name
    file_name = os.path.basename(blob_name)
    # remove the sha256- prefix
    if file_name.startswith("sha256-"):
        file_name = file_name.split("-")[1][:12]

    os.makedirs(os.path.dirname(file_path), exist_ok=True)
    with open(file_path, "wb") as out_file:
        with tqdm.wrapattr(
            out_file,
            "write",
            total=size,
            miniters=1,
            desc=f"pulling {file_name}...",
        ) as file_obj:
            blob.download_to_file(file_obj)


def list_blobs(bucket_name: str, prefix: str) -> list[str]:
    storage_client = storage.Client()
    bucket = storage_client.bucket(bucket_name)
    blobs = bucket.list_blobs(prefix=prefix)
    return [blob.name for blob in blobs]


def blob_exists(bucket_name: str, blob_name: str) -> bool:
    storage_client = storage.Client()
    bucket = storage_client.bucket(bucket_name)
    blob = bucket.blob(blob_name)
    return blob.exists()
