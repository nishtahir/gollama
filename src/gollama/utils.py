import math
import os

__PROJECT_ROOT__ = os.path.abspath(os.path.join(os.path.dirname(__file__), "../.."))


def human_readable_size(size: float, precision: int = 2) -> str:
    if size == 0:
        return "0B"
    size_name = ("B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB")
    i = int(math.floor(math.log(size, 1024)))
    p = math.pow(1024, i)
    s = round(size / p, precision)
    return "%s %s" % (s, size_name[i])
