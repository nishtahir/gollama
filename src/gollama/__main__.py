import os
from typing import Annotated

import timeago
import typer
from rich.console import Console
from rich.table import Table
from typer import Typer

import gollama.gcloud as gcloud
from gollama.model import OllamaManifest, OllamaModel
from gollama.registry import create_registry
from gollama.utils import human_readable_size

app = Typer(no_args_is_help=True)


@app.command("push", help="Push a model to a storage bucket")
def push(
    *,
    bucket_name: Annotated[str, typer.Option(envvar="GCS_BUCKET_NAME")],
    model: Annotated[str, typer.Argument(help="Model to push")],
    ollama_home: Annotated[str, typer.Option(envvar="OLLAMA_HOME")] = "~/.ollama",
    force: Annotated[bool, typer.Option(is_flag=True)] = False,
):
    ollama_home = os.path.expanduser(ollama_home)
    manifests_dir = os.path.join("models", "manifests")
    blobs_dir = os.path.join("models", "blobs")

    ollama_model = OllamaModel.from_name(model)
    manifest_path = os.path.join(manifests_dir, ollama_model.name, ollama_model.version)

    abs_manifest_path = os.path.join(ollama_home, manifest_path)
    manifest = OllamaManifest.from_path(abs_manifest_path)

    print("pushing manifest")
    gcloud.upload_blob(abs_manifest_path, bucket_name, manifest_path, force=force)

    blobs = manifest.blobs()

    for blob in blobs:
        blob_name = blob.replace(":", "-")
        blob_path = os.path.join(blobs_dir, blob_name)
        abs_blob_path = os.path.join(ollama_home, blob_path)
        gcloud.upload_blob(abs_blob_path, bucket_name, blob_path, force=force)

    print("success")


@app.command("list", help="List models in a storage bucket")
def list(
    *,
    bucket_name: Annotated[str, typer.Option(envvar="GCS_BUCKET_NAME")],
):
    registry = create_registry(bucket_name)

    table = Table(box=None)
    table.add_column("NAME")
    table.add_column("SIZE")
    table.add_column("MODIFIED")
    for manifest in registry.manifests:
        table.add_row(
            manifest.name,
            human_readable_size(manifest.size),
            timeago.format(manifest.modified),
        )

    console = Console()
    console.print(table)


@app.command("pull", help="Pull a model from a storage bucket")
def pull(
    *,
    bucket_name: Annotated[str, typer.Option(envvar="GCS_BUCKET_NAME")],
    model: Annotated[str, typer.Argument(help="Model to pull")],
    ollama_home: Annotated[str, typer.Option(envvar="OLLAMA_HOME")] = "~/.ollama",
    force: Annotated[bool, typer.Option(is_flag=True)] = False,
):
    ollama_home = os.path.expanduser(ollama_home)
    manifests_dir = os.path.join("models", "manifests")
    blobs_dir = os.path.join("models", "blobs")

    ollama_model = OllamaModel.from_name(model)
    manifest_path = os.path.join(manifests_dir, ollama_model.name, ollama_model.version)
    abs_manifest_path = os.path.join(ollama_home, manifest_path)

    print("pulling manifest")
    # We always download the manifest, even if it already exists.
    # This is because the manifest is a JSON file, and it may have changed.
    gcloud.download_blob(bucket_name, manifest_path, abs_manifest_path, force=True)
    manifest = OllamaManifest.from_path(abs_manifest_path)

    blobs = manifest.blobs()
    for blob in blobs:
        blob_name = blob.replace(":", "-")
        blob_path = os.path.join(blobs_dir, blob_name)
        abs_blob_path = os.path.join(ollama_home, blob_path)
        gcloud.download_blob(bucket_name, blob_path, abs_blob_path, force=force)


@app.command("rm", help="Remove a model from a storage bucket")
def rm(
    *,
    model: Annotated[str, typer.Argument(help="Model to remove")],
    bucket_name: Annotated[str, typer.Option(envvar="GCS_BUCKET_NAME")],
    ollama_home: Annotated[str, typer.Option(envvar="OLLAMA_HOME")] = "~/.ollama",
):
    ollama_home = os.path.expanduser(ollama_home)
    manifests_dir = os.path.join("models", "manifests")
    blobs_dir = os.path.join("models", "blobs")

    ollama_model = OllamaModel.from_name(model)
    manifest_path = os.path.join(manifests_dir, ollama_model.name, ollama_model.version)
    manifest_content = gcloud.read_blob(bucket_name, manifest_path)
    manifest = OllamaManifest.from_str(manifest_content)

    blobs = manifest.blobs()
    for blob in blobs:
        blob_name = blob.replace(":", "-")
        blob_path = os.path.join(blobs_dir, blob_name)
        gcloud.delete_blob(bucket_name, blob_path)

    gcloud.delete_blob(bucket_name, manifest_path)
    print(f"deleted '{model}'")


if __name__ == "__main__":
    app()
