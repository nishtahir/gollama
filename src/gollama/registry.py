import json
from datetime import datetime

from google.cloud.storage import Blob as GcloudBlob

import gollama.gcloud as gcloud


class Manifest:
    def __init__(self, bucket_name: str, blob: GcloudBlob, text: str):
        self.bucket_name = bucket_name
        self.blob = blob
        self.content = json.loads(text)

    @property
    def name(self) -> str:
        path: str = self.blob.name  # type: ignore
        path = path.removeprefix("models/manifests/")
        path = path.removeprefix("registry.ollama.ai/")
        path = path.removeprefix("library/")

        name, version = path.rsplit("/", 1)
        return f"{name}:{version}"

    @property
    def details(self) -> dict:
        return self.content

    @property
    def modified(self) -> datetime:
        # remove timezone info for timeago
        return self.blob.updated.replace(tzinfo=None)  # type: ignore

    @property
    def size(self) -> int:
        return sum(layer["size"] for layer in self.content["layers"])

    def __str__(self):
        return f"Manifest(blob={self.blob.name})"

    def __repr__(self):
        return self.__str__()


class Blob:
    def __init__(self, blob: GcloudBlob):
        self.blob = blob

    def __str__(self):
        return f"Blob(blob={self.blob.name})"

    def __repr__(self):
        return self.__str__()


class GcloudRegistry:
    def __init__(self, manifests: list[Manifest], blobs: list[Blob]):
        self._manifests = manifests
        self._blobs = blobs

    @property
    def manifests(self) -> list[Manifest]:
        return self._manifests

    @property
    def blobs(self) -> list[Blob]:
        return self._blobs

    def has_blob(self, path: str) -> bool:
        return any(blob.blob.name == path for blob in self.blobs)

    def __str__(self):
        return f"GcloudRegistry(manifests={self.manifests}, blobs={self.blobs})"

    def __repr__(self):
        return self.__str__()


def create_registry(bucket_name: str) -> GcloudRegistry:
    manifest_blobs = gcloud.list_blobs(bucket_name, "models/manifests")
    contents = map(lambda blob: blob.download_as_text(), manifest_blobs)

    iterator = zip(manifest_blobs, contents)
    manifests = [Manifest(bucket_name, blob, content) for blob, content in iterator]

    blobs = [Blob(blob) for blob in gcloud.list_blobs(bucket_name, "models/blobs")]
    return GcloudRegistry(manifests, blobs)
