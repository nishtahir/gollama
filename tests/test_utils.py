from syrupy.assertion import SnapshotAssertion

from gollama.utils import human_readable_size


def test_human_readable_size(snapshot: SnapshotAssertion):
    snapshot.assert_match(human_readable_size(1024))
    snapshot.assert_match(human_readable_size(1024 * 1024))
    snapshot.assert_match(human_readable_size(1024 * 1024 * 1024))
    snapshot.assert_match(human_readable_size(1024 * 1024 * 1024 * 1024))
