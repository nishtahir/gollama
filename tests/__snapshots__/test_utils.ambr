# serializer version: 1
# name: test_human_readable_size
  '1.0 KB'
# ---
# name: test_human_readable_size.1
  '1.0 MB'
# ---
# name: test_human_readable_size.2
  '1.0 GB'
# ---
# name: test_human_readable_size.3
  '1.0 TB'
# ---
